src/devtools/client/debugger/src/components/shared/Modal.tsx

```diff
@@ -10,35 +10,42 @@ import { Transition } from "react-transition-group";
 
 export const transitionTimeout = 50;
 
-export class Modal extends React.Component {
-  onClick = e => {
-    e.stopPropagation();
-  };
+interface IModalProps {
+  additionalClass: string;
+  children: React.ReactNode;
+  handleClose: () => void;
+  status: string;
+  width: string | number;
+}
 
-  render() {
-    const { additionalClass, children, handleClose, status, width } = this.props;
+interface ISlideProps extends IModalProps {
+  in: boolean | undefined;
+}
 
-    return (
-      <div className="modal-wrapper" onClick={handleClose}>
-        <div
-          style={{ width }}
-          className={classnames("modal", additionalClass, status)}
-          onClick={this.onClick}
-        >
-          {children}
-        </div>
+export const Modal = ({ additionalClass, children, handleClose, status, width }: IModalProps) => {
+  const onClick = (e: React.MouseEvent) => e.stopPropagation();
+  return (
+    <div className="modal-wrapper" onClick={handleClose}>
+      <div
+        style={{
+          width,
+        }}
+        className={classnames("modal", additionalClass, status)}
+        onClick={onClick}
+      >
+        {children}
       </div>
-    );
-  }
-}
+    </div>
+  );
+};
 
 export default function Slide({
   width = "50%",
   in: inProp,
   children,
   additionalClass,
   handleClose,
-}) {
+}: ISlideProps) {
   return (
     <Transition in={inProp} timeout={transitionTimeout} appear>
       {status => (
```

Comment:
`Slide` doesn't include `status` in its properties. Where does the reference for status come from here?

Reply:
That looks like a typical "function as a child" setup, where `<Transition>` calls `props.children(status)`

Reply:
So this is a `render callback` pattern? We're relying on `Transition` to run logic when the component mounts and unmounts. However, it's a heavier lift because there's the scheduling of callbacks because of CSS transitions and timeouts. All of this is extracted into `Transition` and the "function as child" pattern allows us to pass in the relevant stateful property and markup we want rendered in an easy-to-use API... and this is a pattern that would normally be a code smell in a codebase but intended for use by library authors. Am I close to understanding this concept?

Reply:
Reading the source code for `react-transition-group` makes me wonder.... this `Transition` component kinda breaks the hooks pattern. I would imagine these actions should be handled in `useLayoutEffect` and there should be some hooks for safely dispatching (and cancelling callbacks). But... this package is maintained by the core React team and is giving people access to lower level APIs like `componentWillEnter`. So I'm guessing there's some more advanced scheduling required for CSS animations/transitions that requires lower-level APIs from React Fiber for scheduling purposes and that's why this library isn't using the "hooks way" of solving this problem. Its been a few years since I've touched React and I'm slightly baffled by this.